ack/utils/image_utils.py

```diff
@@ -0,0 +1,393 @@
+    flipdim: np.ndarry
+        Boolean array informing if the dimensions of the image should be flipped.
+
+    Notes
+    -----
+    The original version of this function can be found at:
+    https://aicsbitbucket.corp.alleninstitute.org/projects/MODEL/repos/image_processing_pipeline/browse/aics_single_cell_pipeline/alignment_tools.py#5
+
+    The docstring for the original version of this function was incorrect.
+    It stated that it took in a CXYZ image but it took in a CYXZ.
+    See `get_features_from_image` for reasoning.
+    """
+    # Get center of mass for the membrane
+    memb_com = proc.get_center_of_mass(proc.get_channel(image, 1))
+
+    # Perform a rigid registration on the image
+    prepped, angle, flipdim = proc.cell_rigid_registration(image)
+
+    # Make sure the nuc and cell channels are binary
+    prepped[0:2] = prepped[0:2] > 0.5
+
+    return prepped, memb_com, angle, flipdim
+
+
+def get_features_from_image(image: np.ndarray) -> Dict:
+    """
+    Generate all segementation, DNA, membrane, and structure features from the provided
+    image.
+
+    Parameters
+    ----------
+    image: np.ndarray
+        The 4D, CYXZ, image numpy ndarray output from
+        `crop_raw_channels_with_segmentation`.
+
+    Returns
+    -------
+    features: Dict
+        A single dictionary filled with features.
+
+    Notes
+    -----
+    The original version of this function can be found at:
+    https://aicsbitbucket.corp.alleninstitute.org/projects/MODEL/repos/image_processing_pipeline/browse/aics_single_cell_pipeline/features.py#8
+
+    The docstring for the original version of this function was incorrect.
+    It stated that it took in a CXYZ image but it took in a CYXZ. This can be seen from
+    line #17 where a transpose to CZYX is done with `img.transpose(0, 3, 1, 2)`.
+    A transpose of (0, 3, 1, 2) on a CXYZ image would result in a CZXY not CZYX.
+    Additionally, simply following the original processing chain shows that the original
+    function is simply handed the output from the original version of the function
+    `crop_raw_channels_with_segmentation` (crop_cell_nuc) which results in a `CYXZ`.
+    """
+    # Get prepared image and feature parameters
+    prepped, memb_com, angle, flipdim = prepare_image_for_feature_extraction(image)
+
+    # Transpose to CZYX
+    prepped = transforms.transpose_to_dims(prepped, "CYXZ", "CZYX")
+
+    # Construct dictionary of basic features
+    regularization_params = {
+        "imsize_orig": image.shape,
+        "com": memb_com.tolist(),
+        "angle": angle,
+        "flipdim": flipdim.tolist(),
+        "imsize_registered": prepped.shape,
+    }
+
+    # Unpack channels
+    nuc_seg = prepped[0]
+    memb_seg = prepped[1]
+    dna_image = prepped[2]
+    memb_image = prepped[3]
+    struct_image = prepped[4]
+
+    # Adjust the DNA and membrane images
+    adjusted_dna_image = ((nuc_seg * dna_image) * 2 ** 8).astype("uint16")
```

Comment:
Do you want to include some kind of commenting on why you're multiplying by 2^8 or is that basic knowledge for someone who is not me?

Reply:
I would love to include some documentation but it is also a mystery to me. Greg originally wrote it. I would _assume_ that feature generation works a bit better with raised intensities as since this is multiplicative the tail of the pixel intensities would be farther skewed making outliers spread farther and easier to get rid of? @donovanr, @jxchen01, and @vianamp may have more insight as well.

Reply:
As far as I know, this is just meant to do the scaling. For this particular code, I think it is better to stick with Greg's original setting, for the sake of reproducing current results. But, in general, this is NOT something very standarized.